common/browser.go

```diff
@@ -464,6 +464,36 @@ func (b *Browser) NewPage(opts goja.Value) api.Page {
 	return browserCtx.NewPage()
 }
 
+// On returns a Promise that is resolved when the browser process is disconnected.
+// The only accepted event value is "disconnected".
+func (b *Browser) On(event string) *goja.Promise {
+	if event != EventBrowserDisconnected {
+		k6Throw(b.ctx, "unknown browser event: %q, must be %q", event, EventBrowserDisconnected)
+	}
+
+	rt := k6common.GetRuntime(b.ctx) //nolint: staticcheck // SA1019 (deprecated)
+	rcb := GetRegisterCallback(b.ctx)
```

Comment:
> I know that abusing the context for this is an anti-pattern

What could be the more challenging and maybe longer way? Any ideas?

> these are functions that will be used extension-wide, and passing them everywhere they're needed would be very cumbersome.

Yeah, but anything that can change behavior is dangerous to put in context.

Reply:
Hhmm the thing is that whatever approach we take with `RegisterCallback`, it should be consistent with what we do for [the other `VU` methods](https://github.com/grafana/k6/blob/a3d7fda089e00bf04ce3d70d4e3cdca43b81e018/js/initcontext.go#L168-L186). Since the context based getters were deprecated in k6, I don't see it as premature optimization, and we should change this soon.

Since these are essentially static within the context of a browser `Launch()`/`Close()`, WDYT if we just attach the `VU` itself to the context? That way we don't have to keep adding methods if/when they change, and can have access to any of the methods throughout the extension?

The problem I see with passing them is that we'll have to change a bunch of signatures, especially with `Runtime()`.  `rg '\.GetRuntime' | wc -l` outputs `158`.

Reply:
> k6common.GetRuntime() was deprecated, but I chose to ignore it for now since we use this everywhere and we'll have to figure out a way of exposing vu.Runtime in a similar way...

This is precisely the problem that we better stop exposing any VU mechanics in the extension code.

> WDYT if we just attach [the VU itself](https://github.com/grafana/xk6-browser/blob/36f44841b3242f1f924eb0b44775b5356b749481/main.go#L71) to the context

I think we shouldn't clutter code with VU mechanics and start refactoring VU-related stuff away as much as we can. So, instead of passing it in a Context, I'm on the side of passing and embedding VU in the types.

> The problem I see with passing them is that we'll have to change many signatures, especially with Runtime(). rg '\.GetRuntime' | wc -l outputs 158.

`Runtime` is most needed because of Goja conversions. If we can provide a translation layer between Goja and the extension code, we can greatly simplify it. For example:

```go
func (b *BrowserContext) WaitForEvent(event string, optsOrPredicate goja.Value) interface{} {
	rt := k6common.GetRuntime(b.ctx)

	if optsOrPredicate != nil && !goja.IsUndefined(optsOrPredicate) && !goja.IsNull(optsOrPredicate) {
		switch optsOrPredicate.ExportType() {
		case reflect.TypeOf(goja.Object{}):
			opts := optsOrPredicate.ToObject(rt)
			for _, k := range opts.Keys() {
				switch k {
				case "predicate":
					predicateFn, isCallable = goja.AssertFunction(opts.Get(k))
					if !isCallable {
						k6Throw(b.ctx, "expected callable predicate")
					}
				case "timeout":
					timeout = time.Duration(opts.Get(k).ToInteger()) * time.Millisecond
				}
			}
```

This kind of cluttering happens because of internally exposing Goja, whereas we could do:

```go
// Extension mechanics package (actual logic here)

package browser

func (c *Context) WaitForEvent(event string, predicate func()) (Event, error) {
	...
	return predicate(c.page)
}

// Translation layer (goja -> extension)

package common

type BrowserContext struct {
	...
	bc browser.Context
	...
}

func (b *BrowserContext) WaitForEvent(event string, optsOrPredicate goja.Value) interface{} {
	// goja stuff here
	r, err := b.bc.WaitForEvent(event predicate)
	if err != nil {
		k6Throw(...)
	}   
}
```

Or something along these lines. But I think this is too much for this discussion right now :) At least, we don't keep cluttering other stuff and get VU-related stuff in an embedded field in each type as a way to pave the future refactoring work. WDYT?

Reply:
Since you think not putting it in the context will be costly, I guess I have a practical idea that combines both approaches:

* Pass it in the context to the `BrowserType`.
* Let it initialize its own VU field.
* Let it have methods that access the VU methods through the VU field.
* Each other type would do the same. They'll extract VU from the context in their `NewX` function. And then fill their own fields.

This can be greatly simplified if we embed VU-accessor in each type.

Reply:
I agree that we should separate the JS layer from our "business" logic, but that's an entirely separate discussion.

I don't entirely follow your last suggestion, but I'll push a change with what I'm thinking and we can discuss it.

Reply:
Let me explain what I mean.

First of all, let's make an internal VU struct:

```go
type VU struct {
	k6modules.VU
	// Makes sense to make it a field since we use it often
	rt *goja.Runtime
}

// WithVU returns a new context based on ctx with the k6 VU instance attached.
func WithVU(ctx context.Context, vu k6modules.VU) context.Context {
	return context.WithValue(ctx, ctxKeyVU, vu)
}

// GetVU returns the attached k6 VU instance from ctx, which can be used to
// retrieve the goja runtime and other k6 objects relevant to the currently
// executing VU.
// See https://github.com/grafana/k6/blob/v0.38.0/js/initcontext.go#L168-L186
func GetVU(ctx context.Context) *VU {
	v := ctx.Value(ctxKeyVU)
	vu, ok := v.(k6modules.VU)
	if !ok {
		return nil
	}

	return &VU{
		VU: vu,
		rt: vu.Runtime(),
	}
}
```

Then:

1. Pass the VU itself in the context (like you did: ✔️):

```go
ctx := common.WithVU(m.vu.Context(), m.vu)

if browserName == "chromium" {
	bt := chromium.NewBrowserType(ctx)
	...
}
```

2. Register the VU in the `Browser`'s field:

```go
type Browser struct {
	...
	vu       *VU
	...
}
```

3. Get VU:

```go
func newBrowser(ctx context.Context, ...) *Browser {
	return &Browser{
		...,
		vu:                  GetVU(ctx),
		...,
	}
}

4. Use VU:

```go
func (b *Browser) On(event string) *goja.Promise {
	...
	cb := b.vu.RegisterCallback()
	p, resolve, reject := b.vu.rt.NewPromise()
	...
}
```

Goal: Not carrying over into the unknown places in the system, for now and the future. And do it in a controlled way.